### ISSUE TITLE
New lint: potentially unwinding caller-supplied operators in `unsafe` code

### ISSUE THREAD

**Issue by Kixunil** [rust-lang/rust-clippy#9200]
### What it does

If a non-`unsafe` generic function containing `unsafe` block uses an operator that is implemented by the caller (via `core::ops::*`) the lint warns that the operator may unwind and suggests `catch_unwind` or calling the method explicitly to signal the programmer thought about it.

Related: #3915
See also: https://github.com/rust-lang/rfcs/pull/3288#discussion_r922871828

### Lint Name

suspicious-operators-in-unsafe

### Category

suspicious

### Advantage

- Forces the programmer to think about unwinds
- Signals to the reviewers "yes, I did think about unwinds" (assuming it's also justified in SAFETY comment)

### Drawbacks

It's quite possible there will be a bunch of FPs. Maybe it'll have to be off by default.

### Example

```rust
impl<T> Foo<T> where T: core::ops::Add {
    pub fn foo(&mut self, val: T) {
        unsafe {
            self.break_invariants();
           // oops: if this unwinds `self` ends up in invalid state
            self.a = self.b + val;
            self.restore_invariants();
        }
    }
}
```

Could be written as:

```rust
impl<T> Foo<T> where T: core::ops::Add {
    pub fn foo(&mut self, val: T) {
        unsafe {
            self.break_invariants();
           // prevent the caller from observing invalid state if `+` unwinds
            let result = catch_unwind(|| self.a = self.b + val);
            self.restore_invariants();
            result.unwrap();
        }
    }
}
```

or:

```rust
impl<T> Foo<T> where T: core::ops::Add {
    pub fn foo(&mut self, val: T) {
        unsafe {
            self.doesnt_break_invariants();
           // safe to call add because the invariants are not broken
            self.a = self.b.add(val).
            self.doesnt_affect_invariants();
        }
    }
}
```

**Comment by ojeda** [comment]
Would the lint need to consider `-Cpanic=abort`?

**Comment by kornelski** [comment]
Libraries typically can't choose their panicking mode (even if a crate had `panic=abort`, the library part could be used as a dependency in a `panic=unwind` binary), so I think it's fair to expect libraries to always consider unwinding.

**Comment by kornelski** [comment]
`panic=abort` has been available since Rust 1.10, and I haven't seen an abort-only library. I guess your concern is related to Rust in Linux environment? But I hope Rust isn't getting fragmented into two dialects that work by different rules.

I do not think that a library that only supports `panic=abort` could be considered *sound* in Rust. This isn't a guarantee that Rust gives. Libraries can't rely on unwinds being always available, but Rust has no guarantee that unwinds will never happen. If I ever saw a library that claims to support only `panic=abort`, I'd file a security advisory for it, because that's an invalid assumption for Rust.

**Comment by kornelski** [comment]
Removal of unwind from `Drop` is insufficient to remove the risks of unwinding that this lint addresses. `Deref` can unwind. `Index` can unwind. All other overloaded operators can unwind. Closures can unwind. I don't think there are any plans to remove all of these other risks.
 
Current Rust semantics are that unwind *may* happen, but is *not guaranteed* to happen. Note that it's different from a guarantee that an unwind will not happen.

Libraries that only support `panic=unwind` are *sound*, even if they don't work as intended, because `panic=abort` will not let unsound code run.

Libraries that only support `panic=abort` risk being *unsound*, because `panic=unwind` may continue to run code and use pointers that are invalid and meant to be stopped by an abort that didn't happen.

> What do you mean by this?

I mean definition what is sound in the language (i.e. whether Rust code is correct) is independent from policies of individual projects. "We only use `panic=abort` setting" doesn't absolve Rust libraries from handling unwinding. 

There's one Rust language, without situational exceptions. Rust doesn't relax meaning `Send` and `Sync` even if you compile for single-threaded WASM. Rust doesn't relax atomics even if you compile for x86 only. And similarly I think Rust shouldn't relax unwind-safety rules even if you compile for use in the Linux kernel.

**Comment by CAD97** [comment]
> There are some supporters of making `-Cpanic=abort` the default, e.g. in the related RFC: Don't allow unwinding from Drop impls rfcs#3288 (comment) @CAD97.

That's a big stronger than I intended; I more meant that while I'm arguing against `-Zpanic-in-drop=abort` being mandatory, I would not argue against `-Cpanic=abort` being the default. It's a (my own) overstatement to say I'm *in favor* of `-Cpanic=abort` being the default. If nothing else, it's a breaking change.

-----

So in general I'm in favor of more allow-by-default `clippy::restriction` lints for controlling unwinds. Warning on operators and silencing by calling the function/method form is clever.

But I also think saying operators unwinding is subtle/invisible the way drop unwinding is subtle/invisible is significantly overstating. Normal use of operators (e.g. `i32 + i32`) can unwind. Calling an operator unwinding isn't an unusual thing; it is a thing that is common and that libraries (should) know that they have to deal with implicitly (if they know about the unwind safety concept) because operators from std unwind under normal conditions.

The btw-operators-can-unwind lint thus shouldn't be limited to downstream types' operators; it should lint against *any* operator which is not known guaranteed to not unwind. (This is not an analysis of the function; this is a property of the signature having `#[nounwind]` or `extern "C"` or other some marker guaranteeing it.)

**Comment by ojeda** [comment]
> Removal of unwind from `Drop` is [...]

Not sure why you bring all this up -- nobody has argued otherwise.

> Libraries that only support `panic=abort` risk being unsound, because panic=unwind may continue to run code and use pointers that are invalid and meant to be stopped by an abort that didn't happen.
>
> I mean definition what is sound in the language (i.e. whether Rust code is correct) is independent from policies of individual projects. "We only use panic=abort setting" doesn't absolve Rust libraries from handling unwinding.

*Correctness* is all about the requirements. If some Rust code is only ever going to be used under a `-Cpanic=abort` setting, and does what is expected of it, it is *correct* even if it does not handle unwinding, because that was not a requirement to begin with.

Rust's *soundness*: if we take the definition from https://doc.rust-lang.org/stable/reference/behavior-considered-undefined.html, then it is unclear how compilation options play with it. If one cannot cause UB with any safe code compiled with the required/documented/intended flags, should we consider it unsound? One could definitely read *"any safe client"* in the widest sense possible, and call it unsound. But such an interpretation may not be too useful for projects that intentionally disable unwinding and may want to design libraries where unwinding does not play a role. At any rate, the definition could be clarified.

But in any case, even if the definition includes all compilation options, one can still create *sound* libraries by simply refusing to compile under the wrong setting:

```rust
#[cfg(not(panic = "abort"))]
compile_error!("Designed for abort-only environments only. Please read the docs!");
```

> There's one Rust language, without situational exceptions. Rust doesn't relax meaning `Send` and `Sync` even if you compile for single-threaded WASM. Rust doesn't relax atomics even if you compile for x86 only. I think Rust shouldn't relax unwind-safety rules even if you compile for use in the Linux kernel.

But nobody is changing the language by refusing to compile, even going with your interpretation of soundness.

>  So "may unwind" is correct, because correct Rust code should be written as if unwinding could happen.

*Sound* Rust code, maybe, depending again on how *soundness* is defined. But *correct*, no.

I understand the worry about splitting the ecosystem, having libraries that cannot be used everywhere etc. Code is not incorrect just because your environment cannot use it.

And if some projects or entities decide it is better for them to write `-Cpanic=abort` code, nobody can really do anything about that.

**Comment by CAD97** [comment]
<details><summary>off-topic</summary>

The terminology I have seen used the most is that a library is e.g. "sound under `-Cpanic=abort`" when relying on certain flags for soundness. A library *is* sound (by the letter of the law, if not the spirit) if it fails to compile when compiled with flags that would make it unsound.

This is in practice the case for more library crates than you'd expect! A *lot* of crates will either refuse to compile or sometimes just be sound with `pointer_width = "16"`. This is generally considered acceptable and falls under the umbrella of the "portability lint;" portability between compiler flags (and not just targets) also falls under the domain of the portability lint just the same.

The difficulty of being unsound with some compiler flags shouldn't be understated, though: if you just `compile_error!` for `panic = "unwind"`, but then a new release of the compiler adds more options (e.g. `panic=unwind-native`, and `panic=unwind` uses a Rust-custom unwinding ABI) then you're unsound until you forbid the new option as well.

For a library to truly be properly sound, it needs to be sound under all build options present and future. If your API is not sound under some present or future option, then you need to gate it to show up only for options which you know are sound.

All of this will hopefully be clearer and easier to talk about if/when the portability lint becomes a thing.

</details>

Especially if this is a allow-by-default opt-in `clippy::restriction` lint (which I strongly think it should be), it doesn't really matter if it saying "this operation may unwind" is technically incorrect under `-Cpanic=abort`, because authors who require `-Cpanic=abort` will just not turn on the lint. Or if they do, they do because they want to know about the unwinding sites because they're auditing the code to relax that requirement.

**Comment by Kixunil** [comment]
@ojeda additional message will bring noise either way. Maybe not too bad but maybe `clippy` could detect `#[cfg(not(panic="abort"))] compile_error!()` (and perhaps also warn if you do it the other way around). The problem with encouraging this practice (and I think this is what @kornelski is trying to say) is that it makes the libraries less useful, potentially forcing people to set `panic = abort`.

OTOH maybe it's OK, I've seen libraries that deliberately don't support 16 bit systems because they handle network messages larger than 2^16 bytes long. If handling unwinds is completely impractical (internal kenel libraries) forcing them to do it is just worse.

**Comment by Kixunil** [comment]
@kornelski I believe @ojeda claims that `compile_error!` behind `cfg` testing for `panic = abort` is sound which seems to match what you say and what I believe to be true. I guess the difference is that @ojeda suggested checking compiler flag while the correct behavior would be to detect `compile_error` behind (correct) cfg.

**Comment by eaglgenes101** [comment]
There exist more than a few clippy lints that are only applicable most of the time, so not including this one on the basis of a possible particular setup that we don't know even exists in practice seems strange.

**Comment by ojeda** [comment]
> I guess the difference is that @ojeda suggested checking compiler flag while the correct behavior would be to detect `compile_error` behind (correct) cfg.

No, I didn't suggest that. In fact, I was the one that suggested the `#[cfg(not(panic = "abort"))] compile_error!` approach...

**Comment by ojeda** [comment]
> adding extra verbosity to the diagnostic

There is no extra verbosity for the common case with what I suggested.

> possible particular set of circumstances that we don't know even exists in practice

`unsafe` code gets compiled under `-Cpanic=abort` all the time and there are abort-only projects out there. I would say chances are quite high that people see this if the lint ends up being enabled by default (which is mentioned as a possibility in OP).

> seems strange

Being less confusing and more correct does not seem like a *"strange"* reason to me.

**Comment by Kixunil** [comment]
@ojeda I mean for clippy to detect that pattern as opposed to flag being set. If that's what you had in mind, then fine, just your repeated `-Cpanic=abort` made me think you want to detect the flag, not the code.

**Comment by ojeda** [comment]
So there is the lint discussion and the soundness discussion (which is why we were all hiding the latter as off-topic). For the former, yes, my suggestion was about the flag (more precisely, the panic strategy of the session), i.e. if `panic=abort`, append something like *"...under `panic=unwind`"* to the message for clarity.
